### ISSUE TITLE
The precision of multi-tab website fingerprinting attacks increases.

### ISSUE THREAD

**Issue by oldersg** [Xinhao-Deng/Website-Fingerprinting-Library#2]
I'm trying to reproduce the DF results for multi-tab website fingerprinting from the ARES paper, but I'm not getting the same performance as reported. Interestingly, I'm seeing an upward trend in precision.
Here are my results:

In the 2-tab setting: {'Precision': 0.7876, 'Recall': 0.5594, 'F1-score': 0.6487}
In the 5-tab setting: {'Precision': 0.8124, 'Recall': 0.1561, 'F1-score': 0.2507}

the experimental parameters I used:

python -u exp/train.py \
  --dataset ${dataset} \
  --model DF \
  --loss BCEWithLogitsLoss \
  --num_tabs 2 (or 5) \
  --device cuda:2 \
  --feature DIR \
  --seq_len 5000 \
  --train_epochs 300 \
  --batch_size 128 \
  --learning_rate 2e-3 \
  --optimizer Adamax \
  --eval_metrics Precision Recall F1-score \
  --save_metric F1-score \
  --save_name max_f1

my setup details:

Python 3.8.19 , torch 2.4.1

Is there anything that needs to be adjusted? Thank you very much.

**Comment by Xinhao-Deng** [comment]
Thank you for your question. Recall and precision exhibit a trade-off relationship, where a decrease in recall often leads to an increase in precision. This occurs because recalling fewer samples reduces the denominator in the precision calculation. We recommend using more common multi-label metrics, such as P@k and AP@k.

The primary differences from the ARES paper can be attributed to three factors:

**Improved Dataset Quality:** We recently developed an image classification model that effectively filters out failed webpage loads by identifying page screenshots, significantly enhancing dataset quality.

**Different Metric Calculations:** The ARES paper evaluates precision using a single-label metric, retaining only one label.

**ARES Paper Did Not Use WFlib:** WFlib was developed after the ARES paper’s publication. There are minor differences between WFlib and the original DF implementation, such as using PyTorch and the save_metric parameter. In our study, DF performance with WFlib is similar to that in the original paper, though not identical.
